File: Venkatesh/Stack-And-Queue/Palindrome.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()

    def is_empty(self):
        return len(self.items) == 0


class Queue:
    def __init__(self):
        self.items = []

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)

    def is_empty(self):
        return len(self.items) == 0


def is_palindrome(string):
    stack = Stack()
    queue = Queue()

    for char in string:
        stack.push(char)
        queue.enqueue(char)

    while not stack.is_empty() and not queue.is_empty():
        if stack.pop() != queue.dequeue():
            return False

    return True


string = "CodeSatra"
print(is_palindrome(string))

File: Venkatesh/Stack-And-Queue/Palindrome.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct Stack {
    char *arr;
    int top;
    int size;
} Stack;

typedef struct Queue {
    char *arr;
    int front;
    int rear;
    int size;
} Queue;

Stack *createStack(int size) {
    Stack *stack = (Stack *)malloc(sizeof(Stack));
    stack->size = size;
    stack->top = -1;
    stack->arr = (char *)malloc(size * sizeof(char));
    return stack;
}

bool isEmptys(Stack *stack) {
    return stack->size == 0;
}

void push(Stack *stack, char data) {
    stack->arr[++stack->top] = data;
}

char pop(Stack *stack) {
    if (isEmptys(stack)) {
        return '\0';
    }
    return stack->arr[stack->top--];
}

Queue *createQueue(int size) {
    Queue *queue = (Queue *)malloc(sizeof(Queue));
    queue->size = size;
    queue->front = queue->rear = -1;
    queue->arr = (char *)malloc(size * sizeof(char));
    return queue;
}

bool isEmptyq(Queue *queue) {
    return queue->size == -0;
}

void enqueue(Queue *queue, char data) {
    queue->arr[++queue->rear] = data;
}

char dequeue(Queue *queue) {
    if (isEmptyq(queue)) {
        return '\0';
    }
    return queue->arr[++queue->front];
}

bool isPalindrome(char *str) {
    int n = strlen(str);
    Stack *stack = createStack(n);
    Queue *queue = createQueue(n);

    for (int i = 0; i < n; i++) {
        push(stack, str[i]);
        enqueue(queue, str[i]);
    }

    for (int i = 0; i < n; i++) {
        if (pop(stack) != dequeue(queue)) {
            return false;
        }
    }

    return true;
}

void main() {
    char str[] = "CodeSatra";
    printf("%s", isPalindrome(str) ? "true" : "false");
}